src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,539 @@
+		this(consumerFactory, new KafkaAckCallbackFactory<K, V>(), topics);
+	}
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory,
+			KafkaAckCallbackFactory<K, V> ackCallbackFactory, String... topics) {
+		Assert.notNull(consumerFactory, "'consumerFactory' must not be null");
+		Assert.notNull(ackCallbackFactory, "'ackCallbackFactory' must not be null");
+		this.consumerFactory = consumerFactory;
+		this.ackCallbackFactory = ackCallbackFactory;
+		this.topics = topics;
+		Object maxPoll = consumerFactory.getConfigurationProperties().get(ConsumerConfig.MAX_POLL_RECORDS_CONFIG);
+		if (maxPoll == null || (maxPoll instanceof Number && ((Number) maxPoll).intValue() != 1)
+				|| (maxPoll instanceof String && Integer.parseInt((String) maxPoll) != 1)) {
+			if (this.logger.isWarnEnabled()) {
+				this.logger.warn("It is advisable to set " + ConsumerConfig.MAX_POLL_RECORDS_CONFIG
+					+ " to 1 to avoid having to seek after each record");
+			}
+		}
+	}
+
+	protected String getGroupId() {
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the message converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized Object doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		TopicPartition topicPartition;
+		synchronized (this.consumerMonitor) {
+			Set<TopicPartition> paused = this.consumer.paused();
+			if (paused.size() > 0) {
+				this.consumer.resume(paused);
+			}
+			ConsumerRecords<K, V> records = this.consumer.poll(this.pollTimeout);
+			this.consumer.pause(this.partitions);
```

Comment:
Why do we still pause if we deal only with one record already?
Maybe I just don't understand how `KafkaConsumer` works...
Thanks

Reply:
SCT wants to use this to run batch jobs. By default, Kafka will rebalance if we don't poll every 5 minutes. `pause` eliminates that problem.